### PR TITLE
reset cache after generation

### LINKED ISSUE #77: Caches should not persist across multiple generate.
When running generate function twice on the same method, the cache on first generation need to be teared down before another generation. Otherwise, we get error like below.

```python
/content/lit-parrot/lit_parrot/model.py in forward(self, x, rope, mask, max_seq_length, input_pos, kv_cache)
    205 
    206         # efficient attention using Flash Attention CUDA kernels
--> 207         y = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, dropout_p=0.0, scale=1.0 / math.sqrt(head_size))
    208 
    209         y = y.transpose(1, 2).contiguous().view(B, T, C)  # re-assemble all head outputs side by side

RuntimeError: The size of tensor a (7) must match the size of tensor b (8) at non-singleton dimension 3
```

We can add a context manager in the Model class and put the generation code under it.

```python
class Parrot(Parrot):
    @contextmanager
    def cache(self):
        try:
            yield
        finally:
            self.kv_caches = []
            self.rope_cache = None
            self.mask_cache = None


# inside generate function
...
with model.cache():
    # generate max_new_tokens tokens
    for _ in range(max_new_tokens):
        x = idx.index_select(0, input_pos).view(1, -1)

        # forward
        logits = model(x, max_seq_length, input_pos)
        logits = logits[0, -1] / temperature
...
```

### PR DESCRIPTION
**Pull request by bkiat1123**
fixes #77 

Add reset method to model and call it before returning in generate function.

### CODE REVIEW

**Comment by ArturK-85** [comment]
Hello @bkiat1123 so I thought maybe to call this function cache_reset

File: generate.py [hunk 1]
```diff
@@ -84,6 +84,7 @@ def generate(
         if idx_next == eos_id:
             return idx[:input_pos]  # include the EOS token
 
+    model.reset()
```
Comment (gkroiz) [inline generate.py:87]: I don't think the reset function should be in the generate function. Maybe move this to `main()`

Comment (bkiat1123) [reply generate.py]: I feels like it is the clean up action of the generate function. We need to clean up after generation, so the next generation start from clean slate. 

Putting outside of generate function risks forgetting to call the reset function, when the generate function is imported to be used somewhere. For example, in the generate_adapter.py script.

Anyway, I have moved it to the main function. Let me know what you think.

Comment (ArturK-85) [reply generate.py]: maybe just add instructions to how_to/inference.md describing the new function

**Review by ArturK-85 — COMMENTED** [review]
I think the right place for the cache_reset function is model.py and add instruction to how_to with an explanation of how use it correctly

File: generate.py [hunk 2]
```diff
@@ -139,6 +137,7 @@ def main(
     for i in range(num_samples):
         t0 = time.perf_counter()
         y = generate(model, encoded, max_new_tokens, temperature=temperature, top_k=top_k)
+        model.reset_cache()
```
Comment (carmocca) [inline generate.py:140]: Reset should be after `t`

File: generate_adapter.py [hunk 3]
```diff
@@ -78,6 +78,7 @@ def main(
     y = generate(
         model, idx=encoded, max_new_tokens=max_new_tokens, temperature=temperature, top_k=top_k, eos_id=tokenizer.eos_id
     )
+    model.reset_cache()
```
Comment (carmocca) [inline generate_adapter.py:81]: Same suggestion

File: lit_parrot/model.py [hunk 4]
```diff
@@ -49,6 +49,11 @@ def _init_weights(self, module: nn.Module) -> None:
             # https://huggingface.co/stabilityai/stablelm-base-alpha-3b/blob/main/config.json#L12
             module.eps = 1e-5
 
+    def reset_cache(self) -> None:
+        self.rope_cache = None
+        self.mask_cache = None
```
Comment (carmocca) [inline lit_parrot/model.py:54]: Why reset these 2? They don't depend on the batch size or input sequence length

File: generate.py (hunk 1 above)
Comment (carmocca) [reply generate.py]: I don't think we need to add this to the howtos, as this is an implementation detail of the code. Maybe just add a comment for why we reset

Comment (carmocca) [reply generate.py]: But I agree that reset should be part of the `main` and not `generate` or the model

Comment (gkroiz) [reply generate.py]: The rational for resetting the cache outside of generate is it is not part of generating text, so it shouldn't be part of the generate function. When benchmarking the times to generate responses, those benchmarks shouldn't include the time to reset cache afterwards.

File: lit_parrot/model.py (hunk 4 above)
Comment (bkiat1123) [reply lit_parrot/model.py]: Right, and block size doesn't change across generate, I removed them.

**Review by carmocca — APPROVED** [review]

**Comment by gkroiz** [comment]
At least on TPUs, I'm noticing a performance discrepancy if we include resetting the rope and mask cache in `reset_cache()`:
Using current reset_cache():
Sample 1: 327.364 ms/token
Sample 2: 208.883 ms/token
Sample 3: 64.096 ms/token

Adding rope and mask resets:
Sample 1: 325.705 ms/token
Sample 2: 64.403 ms/token
Sample 3: 63:039 ms/token

From these results, I suggest we add reseting mask and rope caches? What are your thoughts @carmocca

**Comment by carmocca** [comment]
@gkroiz Why would that be? Also, our timing code does not include the reset, so it shouldn't be different.

**Comment by gkroiz** [comment]
Right, the only difference is the second sample. I'm not sure why there is this discrepancy exists, perhaps it's related to the fact that `rope_cache` and `mask_cache` both use `xla` device? @Liyang90 do you have any thoughts here?

**Comment by Liyang90** [comment]
because these code in the `forward`:
```
        if self.rope_cache is None:
            self.rope_cache = self.build_rope_cache(idx)
        if self.mask_cache is None:
            self.mask_cache = self.build_mask_cache(idx)
```
The first iteration of the first sample compiles from the code processing the prompt when these conditions are `True`, compiles a "graph 1", the second and on iterations are doing generation and compile and uses "graph 2". Without resetting the caches, the first iteration of the second sample compiles from the code processing the prompt when these conditions are `False`, and it is a new graph "graph 3" to be compiled. With them reset, it will reuse the same "graph 1" and no need to compile a new graph.

Ideally, these caches can be initialized in `__init__()` instead of `forward()` if they can be made independent to the input. But for now, this is what happened that explains the time difference.

**Comment by gkroiz** [comment]
@carmocca would it be easiest to add a reset for rope and mask caches? I can create a PR if so.

**Comment by carmocca** [comment]
Go ahead. Let's link the comment above for why it's done.

We should only clear them if `device.type == "xla"`, we can use `self` for this.

**Pull request merged**